src/robot/parsing/model/visitor.py

```diff
@@ -14,27 +14,49 @@
 #  limitations under the License.
 
 import ast
+from typing import Any, Callable, Dict, Optional, Type, Union
 
 from .statements import Node
 
 
+class _NotSet:
+    pass
+
+
 class VisitorFinder:
+    __NOT_SET = _NotSet()
+    __cls_cache: Dict[Type[Any], Union[Callable[..., Any], None, _NotSet]]
 
-    def _find_visitor(self, cls):
-        if cls is ast.AST:
+    def __new__(cls, *_args: Any, **_kwargs: Any):  # type: ignore[no-untyped-def]
+        # create cache on class level to avoid creating it for each instance
+        cls.__cls_cache = {}
+        return super().__new__(cls)
```

Comment:
My comment above was written before the code changed. The new version looks even more complicated than the earlier. It seems like it would try to avoid recreating the cache with new instances that I mentioned above. In that case my proposed implementation is quite a bit simpler.

Reply:
simplify this again in my last commit